umap/umap_.py

```diff
@@ -1387,16 +1376,18 @@ def __init__(
         verbose=False,
         unique=False,
     ):
-
         self.n_neighbors = n_neighbors
         self.metric = metric
-        self.metric_kwds = metric_kwds
         self.output_metric = output_metric
-        if output_metric_kwds is not None:
-            self._output_metric_kwds = output_metric_kwds
+        self.target_metric = target_metric
+        if metric_kwds is None:
+            self.metric_kwds = {}
         else:
-            self._output_metric_kwds = {}
-
+            self.metric_kwds = metric_kwds
+        if output_metric_kwds is None:
+            self.output_metric_kwds = {}
+        else:
+            self.output_metric_kwds = output_metric_kwds
```

Comment:
The sklearn API says that the named attributes should *always* match the parameters that were passed in; that was the reason for the odd indirection with private/underscored attributes holding the values that were actually going to be used in computation (paired with the fact that passing (empty) dicts as default parameters is bad). I would like to try and keep that if possible, even if it is a little awkward. It shouldn't be too tricky to thread that through here...

Reply:
Ah, sorry about that, wasn't editing with those guidelines in mind.  Happy to add it back in and make sure I didn't violate them elsewhere.  How strictly do you want the syntax to adhere to them (e.g., it looks like `self._raw_data` probably "should" be `self.raw_data_`)?

Reply:
fixed

Reply:
In theory if the attribute name ends with an underscore it is a property of the model generated by the fitting process that should be meaningful (i.e. is public); I'm not convinced, for instance, that the raw data should be public, so I use the prefix underscore for that and other similar things.